core/modulesource.go

```diff
@@ -245,6 +249,150 @@ func (src *ModuleSource) AutomaticGitignore(ctx context.Context) (*bool, error)
+// using its context directory.
+//
+// The path is relative to the module context directory, even if it's absolute.
+//
+// Improvement for later: use `.gitignore` to excludes paths from the context.
+func (src *ModuleSource) LoadContext(ctx context.Context, dag *dagql.Server, path string, ignore []string) (inst dagql.Instance[*Directory], err error) {
+	// We exclude .git by default because it's not useful and tends to invalidate the cache.
+	excludes := []string{"**/.git"}
+	includes := []string{}
+
+	for _, p := range ignore {
+		if strings.HasPrefix(p, "!") {
+			includes = append(includes, p[1:])
+		} else {
+			excludes = append(excludes, p)
+		}
+	}
+
+	// If path is not absolute, it's relative to the module root directory.
+	if !filepath.IsAbs(path) {
+		moduleRootPath, err := src.SourceRootSubpath()
+		if err != nil {
+			return inst, fmt.Errorf("failed to get source root subpath for relative context path: %w", err)
+		}
+
+		path = filepath.Join(moduleRootPath, path)
+	}
+
+	switch src.Kind {
+	case ModuleSourceKindLocal:
+		bk, err := src.Query.Buildkit(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get buildkit api: %w", err)
+		}
+
+		slog.Debug("moduleSource.LoadContext: loading contextual directory from local source", "path", path, "kind", src.Kind)
+
+		ctxPath, err := src.resolveContextPath(ctx, bk)
+		if err != nil {
+			return inst, fmt.Errorf("failed to resolve context path: %w", err)
+		}
+
+		// Add the context path to the path to load
+		path = filepath.Join(ctxPath, path)
+
+		path, err = filepath.Rel(ctxPath, path)
+		if err != nil {
+			return inst, fmt.Errorf("path should be relative to the context path: %w", err)
+		}
+
+		_, desc, err := bk.LocalImport(ctx, src.Query.Platform().Spec(), path, excludes, includes)
+		if err != nil {
+			return inst, fmt.Errorf("failed to import local module src: %w", err)
+		}
+
+		inst, err = LoadBlob(ctx, dag, desc)
+		if err != nil {
+			return inst, fmt.Errorf("failed to load local module src: %w", err)
+		}
+
+		return inst, nil
+	case ModuleSourceKindGit:
+		slog.Debug("moduleSource.LoadContext: loading contextual directory from git", "path", path, "kind", src.Kind, "repo", src.AsGitSource.Value.HTMLURL())
+
+		// Use the Git context directory.
+		ctxDir := src.AsGitSource.Value.ContextDirectory.Self
+
+		dir, err := ctxDir.Directory(ctx, path)
+		if err != nil {
+			return inst, fmt.Errorf("failed to load contextual directory %q: %w", path, err)
+		}
+
+		loadedDir, err := dir.AsBlob(ctx, dag)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get dir instance: %w", err)
+		}
+
+		return loadedDir, nil
+	default:
+		return inst, fmt.Errorf("unsupported module src kind: %q", src.Kind)
+	}
+}
+
+// resolveContextPaths returns the context path to the .git directory
+// if it exists. Otherwise, it returns the source root directory.
+func (src *ModuleSource) resolveContextPath(ctx context.Context, bk *buildkit.Client) (string, error) {
```

Comment:
This (and `callerHostFindUpContext`) look like direct copies from `core/schema/modulesource.go`. That makes sense, but we should now remove those implementations, and have the schema ones call these ones.

Duplicates are generally not great when avoidable, since in practice they're fiddly to maintain, and they have a higher potential to drift, which can lead to all sorts of exciting bugs.

Reply:
Agreed.